server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2275,7 +2323,7 @@ public void performRecoveryRestart() throws IOException {
      */
     public void resetRecoveryStage() {
         assert routingEntry().recoverySource().getType() == RecoverySource.Type.PEER : "not a peer recovery [" + routingEntry() + "]";
-        assert currentEngineReference.get() == null;
+        assert this.currentEngine == null;
```

Comment:
nit: We're not guarding this assertion with the RW lock when we call this from `RecoveryTarget`

Reply:
Thanks for catching this